crates/wasi-nn/src/backend/onnxruntime.rs

```diff
@@ -0,0 +1,182 @@
+        let inputs = self
+            .0
+            .lock()
+            .unwrap()
+            .inputs
+            .iter()
+            .map(|_| None)
+            .collect::<Vec<_>>();
+        let outputs = self
+            .0
+            .lock()
+            .unwrap()
+            .outputs
+            .iter()
+            .map(|_| None)
+            .collect::<Vec<_>>();
+        let box_: Box<dyn BackendExecutionContext> = Box::new(ONNXExecutionContext {
+            session: self.0.clone(),
+            inputs,
+            outputs,
+        });
+        Ok(box_.into())
+    }
+}
+
+struct ONNXExecutionContext {
+    session: Arc<Mutex<Session>>,
+    inputs: Vec<Option<Tensor>>,
+    outputs: Vec<Option<Vec<u8>>>,
+}
+
+unsafe impl Send for ONNXExecutionContext {}
+unsafe impl Sync for ONNXExecutionContext {}
+
+impl BackendExecutionContext for ONNXExecutionContext {
+    fn set_input(&mut self, index: u32, tensor: &Tensor) -> Result<(), BackendError> {
+        self.inputs[index as usize].replace(tensor.clone());
+        Ok(())
+    }
+
+    fn compute(&mut self) -> Result<(), BackendError> {
+        let shaped_inputs: Vec<_> = self
+            .inputs
+            .iter()
+            .enumerate()
+            .map(|(i, _o)| {
+                let input = self.inputs[i].as_ref().unwrap();
+                let dims = input
+                    .dimensions
+                    .as_slice()
+                    .iter()
+                    .map(|d| *d as usize)
+                    .collect::<Vec<_>>();
+                match input.tensor_type {
+                    TensorType::Fp32 => {
+                        let data = bytes_to_f32_vec(input.data.to_vec());
+                        inputs![Array::from_shape_vec(dims, data).unwrap()].unwrap()
+                    }
+                    TensorType::Fp16 => {
+                        unimplemented!("Fp16 not supported by ONNX");
+                    }
+                    TensorType::Bf16 => {
+                        unimplemented!("Bf16 not supported by ONNX");
+                    }
+                    TensorType::U8 => {
+                        unimplemented!("U8 not supported by ONNX");
+                        // Array::from_shape_vec(dims, input.data.to_vec()).unwrap()
+                    }
+                    TensorType::I32 => {
+                        unimplemented!("I32 not supported by ONNX");
+                        // Array::from_shape_vec(dims, input.data.to_vec()).unwrap()
+                    }
+                    TensorType::I64 => {
+                        unimplemented!("I64 not supported by ONNX");
+                        // Array::from_shape_vec(dims, input.data.to_vec()).unwrap()
+                    }
+                    TensorType::Fp64 => {
+                        unimplemented!("Fp64 not supported by ONNX");
+                        // Array::from_shape_vec(dims, input.data.to_vec()).unwrap()
+                    }
```

Comment:
Maybe something like the following would be less verbose?

```
_ => {
  unimplemented!("{:?} not supported by ONNX", input.tensor_type);
}
```

Reply:
I was verbose to be super clear about what is not implemented. I've since changed it to your suggestion.